### PR TITLE
[WIP] Upgrade to relay v6

### PR DESCRIPTION
**Pull request by alloy**
- [x] Move all local `@types` modifications to upstream. https://github.com/alloy/DefinitelyTyped/tree/stash-relay-work
- [x] Test on our application codebase.

### CODE REVIEW

**Comment by sibelius** [comment]
the current master of  `relay-compiler-language-typescript` was breaking in your app?

**Comment by kastermester** [comment]
I will attempt to get this tested on our codebase during the week :)

What exactly is missing for `@match` support?

**Comment by alloy** [comment]
> the current master of `relay-compiler-language-typescript` was breaking in your app?

Nah. I thought so at first, but that ended up being related to some other local patches we have for Relay.

**Comment by alloy** [comment]
> I will attempt to get this tested on our codebase during the week :)

Ace 👍 

> What exactly is missing for `@match` support?

Mostly just needing to re-organize code so that an object typing can be extended with multiple props, rather then a single. The flow code uses a spread, which TS doesn’t support. So we'd either have to iterate over all the props and add them one by one or do an intersection with the object type. Just something I didn’t want to get bogged down by right now as I don’t expect anybody to really be using this–definitely not us.

**Comment by alloy** [comment]
@kastermester I'm going to release this as-is, it seems to work well on our codebase at least. Would still love to hear how it works for you, though!

File: src/TypeScriptGenerator.ts [hunk 1]
```diff
@@ -490,6 +524,101 @@ function createVisitor(options: TypeGeneratorOptions) {
...
+function makeRawResponseProp(
+  { key, schemaName, value, conditional, nodeType, nodeSelections }: Selection,
+  state: State,
+  concreteType: string | null
+) {
+  if (nodeType) {
+    if (nodeType === MODULE_IMPORT_FIELD) {
+      // TODO: In flow one can extend an object type with spread, with TS we need an intersection (&)
+      // return ts.createSpread(ts.createIdentifier(key));
+      throw new Error("TODO!");
```
Comment (alloy) [inline src/TypeScriptGenerator.ts]: @kastermester FYI, here’s that `@match` todo

Comment (kastermester) [reply src/TypeScriptGenerator.ts]: Great - I might take a stab at this at some point. Now that I read that I do recall this from last time I ported the code.

**Comment by kastermester** [comment]
As far as I have been able to tell the only difference (for us) is that in the old type generation sometimes some extra fields were emitted erroneously before. Which are now gone. A fragment like this:

```graphql
fragment LikeCounter_reactionable on Reactionable
	@argumentDefinitions(size: { type: "Int!", defaultValue: 1 }, cursor: { type: "String" }) {
	id
	reactions(reactionTypes: [Like]) {
		allProfiles(first: $size, after: $cursor) @connection(key: "LikeCounter_allProfiles") {
			total
			pageInfo {
				endCursor
				hasNextPage
			}
			edges {
				node {
					id
					databaseId
					name
					picture {
						thumbnailURL: thumbnailURL(type: "web")
					}
					...ProfileFallbackImage_profile
				}
			}
		}
		allProfilesTotal
		reactionable {
			id
		}
	}
	viewerReaction
}
```

Generates the following diff:

```diff
                     readonly picture: {
                         readonly thumbnailURL: string;
                     } | null;
-                    readonly __typename: string;
                     readonly " $fragmentRefs": ProfileFallbackImage_profile$ref;
                 };
-                readonly cursor: string;
             }>;
         };
```

So this just seems like a huge bugfix where the old type generation was in the wrong. Also there's some changes around how a field selected with its default argument specified gets transformed in the query - but all seems to work as intended. Very nice! :)

**Comment by alloy** [comment]
Yup, I should have explicitly mentioned that, but indeed those erroneous typings are now gone ✨

**Pull request merged**